core-plugin/src/main/groovy/com/github/jrubygradle/api/gems/GemVersion.groovy

```diff
@@ -0,0 +1,573 @@
+package com.github.jrubygradle.api.gems
+
+import groovy.transform.CompileDynamic
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+
+import java.util.regex.MatchResult
+import java.util.regex.Pattern
+
+import static com.github.jrubygradle.api.gems.GemVersion.Boundary.EXCLUSIVE
+import static com.github.jrubygradle.api.gems.GemVersion.Boundary.INCLUSIVE
+import static com.github.jrubygradle.api.gems.GemVersion.Boundary.OPEN_ENDED
+
+/**
+ * With rubygems almost all dependencies will be declared
+ * via versions ranges and tools like Bundler are very strict on how to
+ * resolve those versions - i.e. the resolved version needs to obey each given
+ * constraint. Ivy does the same but Gradle and Ivy pick the latest and
+ * newest version when there are more then one constraint for the same gem -
+ * which can create problems when using Bundler alongside Gradle.
+ *
+ * When converting a GemSpec into a Ivy ivy.xml the translation of a
+ * gem version range into an Ivy version range. typically '~> 1.0' from ruby
+ * becomes {@code [1.0.0,2.0[} on the Ivy side. so most dependencies from
+ * gem artifacts will use such version ranges.
+ *
+ * To help gradle to be closer to the rubygems world when resolving gem
+ * artifacts, it needs to calculate intersection between version ranges
+ * in maven manner.
+ *
+ * This class basically represents an Ivy version range with boundary
+ * (exclusive vs. inclusive or open-ended) and its lower and upper bounded version and
+ * allows to intersect its range with another version range.
+ *
+ * It also translate fixed version '1.0' to [1.0, 1.0] or the gradle notation
+ * 1.2+ to [1.2, 1.99999] or 1.+ to [1.0, 1.99999] following the gemspec-to-pom
+ * pattern.
+ *
+ * @author Christian Meier
+ * @author Schalk W. Cronjé
+ *
+ * @since 2.0 (Moved here from base plugin where it existed since 0.4.0)
+ */
+@CompileStatic
+@Slf4j
+class GemVersion implements Comparable<GemVersion> {
```

Comment:
I kept a lot of the core logic the same, but reworked parts to be closer to an Ivy setup rather than Maven. I also used an enumeration to keep track of boundaries rather than using strings. Also made it distinguish between open boundaries and not just exclusive/inclusive boundaries.